### PR TITLE
feat: argument based prefetcher

### PR DESCRIPTION
**Pull request by jakmeier**
Prefetcher that loads IO requests in parallel if the keys are listed in the  smart contract function call argument.

This is a experimental feature that has proven to stabilize mainnet nodes under current traffic. Especially when the node is not running with fast local SSDs.

In the future we should generalize this implementation.

### CODE REVIEW

File: runtime/runtime/src/prefetch.rs [hunk 1]
```diff
@@ -70,10 +87,46 @@ impl TriePrefetcher {
...
+                                self.prefetch_sweat_record_batch(
+                                    account_id.clone(),
+                                    &fn_call.args,
+                                )?;
+                            }
+                        }
+                    }
+                }
+
+                // SWEAT specific argument prefetcher
```
Comment (mm-near) [inline runtime/runtime/src/prefetch.rs:115]: code repeated

**Review by mm-near — APPROVED** [review]

File: core/store/src/trie/prefetching_trie_storage.rs [hunk 2]
```diff
@@ -63,6 +68,14 @@ pub struct PrefetchApi {
     /// to mark what is already being fetched, to avoid fetching the same data
     /// multiple times.
     pub(crate) prefetching: PrefetchStagingArea,
+
+    pub enable_receipt_prefetching: bool,
+    /// Configured accounts will be prefetched as SWEAT token account, if predecessor is listed as sender.
```
Comment (nikurt) [inline core/store/src/trie/prefetching_trie_storage.rs:73]: `if predecessor is listed as sender` sounds confusing.

File: core/store/src/trie/prefetching_trie_storage.rs [hunk 3]
```diff
@@ -249,6 +285,17 @@ impl TriePrefetchingStorage {
...
 impl PrefetchStagingArea {
+    fn new(shard_id: ShardId) -> Self {
+        let inner = InnerPrefetchStagingArea {
+            slots: SizeTrackedHashMap {
+                map: Default::default(),
+                size_bytes: 0,
+                metrics: StagedMetrics::new(shard_id),
+            },
+        };
+        Self(Arc::new(Mutex::new(inner)))
```
Comment (nikurt) [inline core/store/src/trie/prefetching_trie_storage.rs:296]: Should `slots.update_metrics()` be called upon struct creation?

File: core/store/src/trie/shard_tries.rs [hunk 4]
```diff
@@ -145,7 +151,6 @@ impl ShardTries {
             block_hash,
             is_view,
         );
-
```
Comment (Ekleog-NEAR) [inline core/store/src/trie/shard_tries.rs:148]: You’ll have less merge conflicts without that change

File: runtime/runtime/src/prefetch.rs [hunk 5]
```diff
@@ -121,12 +160,44 @@ impl TriePrefetcher {
...
             debug!(target: "prefetcher", "I/O scheduler input queue full, dropping prefetch request");
             Err(())
         } else {
+            self.prefetch_enqueued.inc();
             Ok(())
         }
     }
+
+    /// Prefetcher specifically tuned for SWEAT record batch
+    fn prefetch_sweat_record_batch(&self, account_id: AccountId, arg: &[u8]) -> Result<(), ()> {
```
Comment (nikurt) [inline runtime/runtime/src/prefetch.rs:173]: Please comment that this function is temporary and add a link to the flat-storage github issue.

**Review by nikurt — APPROVED** [review]

File: runtime/runtime/src/prefetch.rs [hunk 6]
```diff
@@ -121,12 +160,44 @@ impl TriePrefetcher {
...
+            if json.is_object() {
+                if let Some(list) = json.get("steps_batch") {
+                    if let Some(list) = list.as_array() {
+                        for tuple in list.iter() {
+                            if let Some(tuple) = tuple.as_array() {
+                                if let Some(user_account) = tuple.first().and_then(|a| a.as_str()) {
+                                    let hashed_account =
+                                        sha2::Sha256::digest(user_account.as_bytes()).into_iter();
+                                    let sweatcoin_prefix = [0x74, 0x00];
+                                    let mut key = sweatcoin_prefix.to_vec();
```
Comment (Ekleog-NEAR) [inline runtime/runtime/src/prefetch.rs:184]:
```suggestion
                                    let mut key = vec![0x74, 0x00];
```

**Review by Ekleog-NEAR — APPROVED** [review]

File: core/store/src/trie/prefetching_trie_storage.rs (hunk 3 above)
Comment (jakmeier) [reply core/store/src/trie/prefetching_trie_storage.rs:297]: done

File: runtime/runtime/src/prefetch.rs (hunk 5 above)
Comment (jakmeier) [reply runtime/runtime/src/prefetch.rs:176]: done

**Pull request merged**